File: c/binarySearch.c
/*
Binary Search algo Implementation: 

Repeatedly divides the search interval in half
*/

# include<stdio.h>

int binarySearch(int arr[], int start, int end, int key){
    while(start <= end){
        int mid = (start + end) / 2;
        if(arr[mid] == key){
            return mid;
        }
        else if(arr[mid] < key){
            start = mid + 1;
        }
        else{
            end = mid - 1;
        }
    }
    return -1;
}

int main(){
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int sizeOfArray = sizeof(arr)/sizeof(arr[0]);
    int key;
    printf("Enter the key to be searched: ");
    scanf("%d", &key);
    int result = binarySearch(arr, 0, sizeOfArray - 1, key);
    if (result == -1){
        printf("Key not found\n");
    }
    else{
        printf("Key found at index %d\n", result);
    }
    return 0;
}

File: rust/src/binarySearch.rs
fn binary_search(arr: &[i32], mut start: i32, mut end: i32, target: i32) -> Option<usize> {
    while start <= end {
        let mid = (start + end) / 2;
        let mid_index = mid as usize;  // Convert mid to usize for array indexing

        if arr[mid_index] == target {
            return Some(mid_index);
        } else if arr[mid_index] > target {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }
    None
}

fn main() {
    let mut arr = vec![10, 23, 45, 70, 11, 15];
    arr.sort();  // Ensure the array is sorted for binary search
    let target = 45;

    match binary_search(&arr, 0, arr.len() as i32 - 1, target) {
        Some(index) => println!("Element found at index {}", index),
        None => println!("Element not found"),
    }
}
